### ISSUE TITLE
Padded Tensor should not be full '-inf' In Spatial Encoder for Graphormer

### ISSUE THREAD

**Issue by ArthurLeoM** [dmlc/dgl#5476]
## Original source code and the problem

In Graphormer, the graph topology is encoded by the Spatial Encoder, used as the attention bias in GraphormerLayer. However, when conducting batch training, the graphs are in different sizes and need to conduct padding.
At Line 541-546 from the source code of ```nn.pytorch.graph_transformer.py```, we can see that we pad '-inf' for all the empty entries.
```
padded_encoding = th.full(
  (max_num_nodes, max_num_nodes, self.num_heads), float("-inf")
).to(device)

padded_encoding[0:num_nodes, 0:num_nodes] = dist_embedding
spatial_encoding.append(padded_encoding)
```
However, for graphs whose nodes are less than ```max_num_nodes```, the rows after ```num_nodes``` will be full of '-inf', and when passed to the argument ```attn_bias``` in ```GraphormerLayer```, the weights will be softmaxed and the attention generated will be full 'nan' for those rows. This can make the output of the GraphormerLayer also full of 'nan' for those empty nodes. Further in the next layer, the normal nodes will query the empty nodes with attention weight 0 (i.e., 'nan' will be multiplied by attention weights 0) and generate 'nan' outputs for all the nodes. This is obviously a bad behavior.

## Solution

We can calculate the shortest path for the empty nodes, which means to set the diagonal values of ```padded_conding``` to zeros. Here is my implementation (modified Line 541-546 in ```graph_transformer.py```):
```
padded_encoding = th.full(
    (max_num_nodes, max_num_nodes, self.num_heads), float("-inf")
).to(device)
for i in range(max_num_nodes):
    padded_encoding[i, i] = 0.
padded_encoding[0:num_nodes, 0:num_nodes] = dist_embedding
spatial_encoding.append(padded_encoding)
```

This can solve the 'nan' problem, because the softmax will generate weight 1 for the non-inf entry. Moreover, the attention weights for the empty nodes will still be zeros and not be queried.

## Environment

 - DGL Version (e.g., 1.0): 1.0.1+cu113
 - Backend Library & Version : PyTorch 1.11.0
 - OS: Ubuntu
 - How you installed DGL: conda
 - Python version: 3.9.12
 - CUDA/cuDNN version (if applicable): cuda 11.4
 - GPU models and configuration: Nvidia RTX 3090

**Comment by jermainewang** [comment]
Hi @ArthurLeoM , thanks for reporting the issue! The `nn.pytorch.graph_transformer.py` is still under development and you've seem to caught a potential bug already 😄 . Please use them with caution.

I'm cc'ing @rudongyu for visibility. Dongyu, can you confirm whether this is a bug or not?

**Comment by rudongyu** [comment]
Hi @AuthurLeoM, thanks for raising the issue! This issue is expected to be handled by later [computation](https://github.com/dmlc/dgl/blob/master/python/dgl/nn/pytorch/graph_transformer.py#L331) of BiasedMultiheadAttention in our current design.

After all kinds of `attn_bias` are applied, some attention weights are set to -inf according to input `attn_mask` of BiasedMultiheadAttention: `attn_weights[attn_mask.to(th.bool)] = float("-inf")`. Carefully setting `attn_mask` avoids potential 'nan' pollution to valid nodes. Rows for unexisting nodes should be full of `False`. (The initialization of all attn_bias can be zero.)

Sorry for the confusion, we will add more description on the parameter `attn_mask` in docs and modify the initialization of attn_bias to zero tensors.

**Comment by ArthurLeoM** [comment]
Hi @rudongyu , thanks for your response! I also noticed the behavior of ```attn_weights```. However, I think it is also questioned to set all rows of unexisting nodes to ```False```, because we calculate attention weights (before softmax) with ```attn_weights[attn_mask.to(th.bool)] = float("-inf")```, while the Softmax result of a tensor full of ```-inf``` will be full of ```nan```s. Therefore, this kind of pollution will still exist.
I think it might be an effective solution to set the diagonal values of the ```attn_mask``` matrix to 0. With this setting, the unexisting nodes will only interact with themselves (the attention weights will be 0 for other nodes and 1 for itself), and will not lead to the pollution phenomenon.

**Comment by rudongyu** [comment]
@ArthurLeoM, thanks for sharing your opinion! Setting diagonal values to 0 is indeed a reasonable solution. In fact, any solution that leaves more than 1 value unmasked with -inf for rows of unexisting nodes is valid. Besides, please note that a `True` value in `attn_mask` indicates that the corresponding position is not allowed to attend, following the convention of MultiheadAttention module in Pytorch. We leave the construction of `attn_mask` to dataloading part. I would consider initializing it with `torch.zeros(bsz, max_N, max_N, dtype=torch.bool)`,  then masking invalid positions of i-th graph with `attn_mask[i, :, N:] = True`, which seems simple and straightforward.

**Comment by zdhNarsil** [comment]
The loop
```python
for i in range(max_num_nodes):
    padded_encoding[i, i] = 0.
``` 
could be optimized with 
```python
torch.diagonal(padded_encoding).fill_(0.)
``` 
@ArthurLeoM
